rfc/2024-11-13-provider-oci-registry.md

```diff
@@ -0,0 +1,240 @@
+
+</summary>
+
+As a provider author, you may want to package up your provider binary into an OCI image. This involves the following steps:
+
+1. Take all build artifacts and add them to a container image.
+2. Upload the container image to an OCI registry.
+
+For example, assuming you have your binary built and wanted to package it up into a container, you could use the following `Dockerfile`:
+
+```Dockerfile
+FROM scratch
+
+ADD terraform-provider-aws
+ADD README.md
+ADD LICENSE
+ADD CHANGELOG.md
+```
+
+You could now run `docker build -t ghcr.io/opentofu/terraform-provider-aws:1.2.3-amd64 .` followed by `docker push ghcr.io/opentofu/terraform-provider-aws:1.2.3`.
+
+For more advanced use cases, you could enable container image building in your [goreleaser config](https://goreleaser.com/customization/docker/) to enable building multi-arch images.
+</details>
+
+<details>
+<summary>
+
+#### 📂 ORAS approach
+
+</summary>
+
+As a provider author, you may want to package up your provider binary into an OCI image. You can do this by using the [ORAS command line tool](https://oras.land/docs/installation) using the following commands:
+
+```
+oras login -u username -p password ghcr.io
+oras push ghcr.io/opentofu/terraform-provider-aws:1.2.3 ./dist:application/octet-stream+tar
+```
+
+🛑🛑🛑 TODO: how do you create multi-arch with ORAS? 🛑🛑🛑
+
+</details>
+
+### Mirror configuration
+
+If you would like to use an OCI registry as your OpenTofu mirror, you will be able to do so by specifying a `provider_installation` block with an `oci_mirror` section as follows:
+
+```hcl
+provider_installation {
+  oci_mirror {
+    url = "ghcr.io/{{ .Namespace }}/terraform-provider-{{ .Name }}"
+  }
+}
+```
+
+Note that the `url` field is a Go template to resolve the container image address containing the provider binary. You can use `{{ .Namespace }}` to add the provider namespace (e.g. `hashicorp`) and `{{ .Name }}` to add the provider name (e.g. `aws`) to the container image address.
+
+> [!NOTE]
+> OpenTofu will **not** start a container from this image, but instead use the container image as a way to extract the provider binary. Also note that providers are not automatically available in any particular container registry. This feature is intended for advanced use cases.
+
+### The first-class configuration
+
+If you would like to use a provider hosted in an OCI registry, you may do so by invoking a well-known public registry in your `required_providers` block.
+
+```hcl
+terraform {
+  required_providers {
+    aws = {
+      source  = "ghcr.io/opentofu/terraform-provider-aws"
+      version = "~> 1.0"
+    }
+  }
+}
+```
+
+OpenTofu automatically identifies the following registries as OCI registries:
+
+- `docker.io`
+- `ghcr.io`
+- `containers.pkg.github.com`
+- `quay.io`
+- `gcr.io`
+- `public.ecr.aws`
+
+🛑🛑🛑 TODO: How does a user specify a custom registry? Do we perform some sort of service discovery? Do we distribute a list of known registries? 🛑🛑🛑
```

Comment:
Building on my previous comment above, where I said "we could still offer an explicit service-discovery-based approach", I think the most general case is to use our existing service discovery protocol and define a new service identifier, such as `oci-providers.v1`, whose value is a template similar to the one we discussed for the `oci_mirror` block above.

The "magic domain name" I mentioned in the previous comment would (I think) allow most folks to avoid setting up any OpenTofu-specific service discovery, but offering service discovery as the most general solution means that folks would have the full flexibility to set up something custom if our default mapping doesn't work for them.

We don't yet have any precedent for the discovery results in the service discovery document being treated as templates rather than just static strings, but we could potentially follow a similar approach as I proposed for the CLI configuration above and use HCL templates in there:

```json
{
  "oci-providers.v1": "oci.example.com/${namespace}/tf-${name}"
}
```

We'd need to make a somewhat-arbitrary decision about which service "wins" if the service discovery document includes both `providers.v1` and `oci-providers.v1`. That seems pretty unlikely to occur in practice, so I don't think it really matters which preference order we choose as long as we document it.

(Note that the service discovery document doesn't actually need to be hosted on the same hostname where the OCI registry is running, but it would need to be published on whatever hostname the organization intends to use as the source address part of their provider source addresses.)